File: modules/core/src/main/java/org/openlmis/core/domain/ProgramProduct.java
/*
 * This program is part of the OpenLMIS logistics management information system platform software.
 * Copyright © 2013 VillageReach
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *  
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.
 * You should have received a copy of the GNU Affero General Public License along with this program.  If not, see http://www.gnu.org/licenses.  For additional information contact info@OpenLMIS.org. 
 */

package org.openlmis.core.domain;

import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import org.openlmis.core.exception.DataException;
import org.openlmis.upload.Importable;
import org.openlmis.upload.annotation.ImportField;

import static com.fasterxml.jackson.databind.annotation.JsonSerialize.Inclusion.NON_EMPTY;

/**
 * ProgramProduct represents a product available under a program and program-product specific attributes like currentPrice and dosesPerMonth.
 * Also defines contract for upload of a ProgramProduct mapping.
 */
@Data
@NoArgsConstructor
@EqualsAndHashCode(callSuper = false)
@JsonIgnoreProperties(ignoreUnknown = true)
@JsonSerialize(include = NON_EMPTY)
public class ProgramProduct extends BaseModel implements Importable {

  @ImportField(name = "Program Code", type = "String", nested = "code", mandatory = true)
  private Program program;

  @ImportField(name = "Product Code", type = "String", nested = "code", mandatory = true)
  private Product product;

  @ImportField(name = "Doses Per Month", type = "int", mandatory = true)
  private Integer dosesPerMonth;

  @ImportField(name = "Is Active", type = "boolean", mandatory = true)
  private Boolean active;

  @ImportField(mandatory = true, type = "String", name = "Product Category", nested = "code")
  private ProductCategory productCategory;

  private Long productCategoryId;

  @ImportField(name = "Full Supply", type = "boolean", mandatory = true)
  private boolean fullSupply;

  @ImportField(name = "Display Order", type = "int")
  private Integer displayOrder;

  ProgramProductISA programProductIsa;

  private Money currentPrice;

  public ProgramProduct(Long id) {
    super(id);
  }

  public ProgramProduct(Program program, Product product, Integer dosesPerMonth, Boolean active) {
    this.program = program;
    this.product = product;
    this.dosesPerMonth = dosesPerMonth;
    this.active = active;
  }

  public ProgramProduct(Program program, Product product, Integer dosesPerMonth, Boolean active, Money currentPrice) {
    this.program = program;
    this.product = product;
    this.dosesPerMonth = dosesPerMonth;
    this.active = active;
    this.currentPrice = currentPrice;
  }

  public void validate() {
    if (currentPrice.isNegative()) throw new DataException("programProduct.invalid.current.price");
  }

  public ProgramProduct(ProgramProduct programProduct) {
    this.id = programProduct.id;
    this.program = programProduct.program;
    this.product = programProduct.product;
    this.dosesPerMonth = programProduct.dosesPerMonth;
    this.active = programProduct.active;
    this.currentPrice = programProduct.currentPrice;
    this.programProductIsa = programProduct.programProductIsa;
    this.displayOrder = programProduct.displayOrder;
    this.fullSupply = programProduct.fullSupply;
  }
}

File: modules/openlmis-web/src/main/webapp/public/js/program-product/model/program-product.js
/*
 * This program is part of the OpenLMIS logistics management information system platform software.
 * Copyright © 2013 VillageReach
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *  
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.
 * You should have received a copy of the GNU Affero General Public License along with this program.  If not, see http://www.gnu.org/licenses.  For additional information contact info@OpenLMIS.org. 
 */

var ProgramProduct = function (programProduct) {
  $.extend(true, this, programProduct);

  ProgramProduct.prototype.calculateISA = function (facility, period) {
    if (isUndefined(programProduct.programProductIsa) && isUndefined(programProduct.overriddenIsa)) {
      programProduct.isaAmount = "--";
    } else {
      programProduct.programProductIsa = new ProgramProductISA(programProduct.programProductIsa);
      programProduct.isaAmount = (!isUndefined(programProduct.overriddenIsa)) ? programProduct.overriddenIsa : programProduct.programProductIsa.calculate(facility.catchmentPopulation);
//          TODO important need validation on packSize to be more than 0
      programProduct.isaAmount = programProduct.isaAmount ? Math.ceil((programProduct.isaAmount * period.numberOfMonths) / programProduct.product.packSize) : 0;
    }

    this.isaAmount = programProduct.isaAmount;
  };


};

ProgramProduct.groupProductsMapByName = function (facility, otherGroupName) {
  var programProductsMap = _.groupBy(facility.supportedPrograms[0].programProducts, function (programProduct) {
    return programProduct.product.productGroup ? programProduct.product.productGroup.name : otherGroupName;
  });
  return programProductsMap;
};

ProgramProduct.calculateProductIsaTotal = function (aggregateProduct, productTotal) {
  if (!isNaN(utils.parseIntWithBaseTen(aggregateProduct.isaAmount))) {
    if (productTotal.isaAmount == "--") {
      productTotal.isaAmount = aggregateProduct.isaAmount;
    } else {
      productTotal.isaAmount = productTotal.isaAmount + aggregateProduct.isaAmount;
    }
  }
};